### PR TITLE
nixos/xdg/xdg-open: init

### PR DESCRIPTION
**Pull request by LunNova**
###### Description of changes

Potential fix for #160923 although is not currently enabled by default.

MERGED ~~Until https://github.com/NixOS/nixpkgs/pull/181171 gets merged, this relies on enabling an overlay for all those packages to actually use the xdg-open implementation from the path.~~

~~Many other packages rely on it implicitly without adding it in a wrapper so will work already without the overlay.~~

Some other things like the steam FHS env also require the overlay currently. Could get them to pick it up from the PATH in a subsequent PR?

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/config/xdg/xdg-open.nix [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  options.xdg.xdg-open = {
+    enable = lib.mkEnableOption "adding an xdg-open implementation to the path (environment.systemPackages)";
+    implementation = lib.mkOption {
+      description = "Which implementation to use";
+      type = lib.types.enum (builtins.attrNames implementations);
+      # This makes the most sense as a default for NixOS
+      # due to https://github.com/NixOS/nixpkgs/issues/160923
+      default = "xdg-open-with-portal";
+    };
+    # TODO:
```
Comment (LunNova) [inline nixos/modules/config/xdg/xdg-open.nix:27]: - [ ] FIXME: needs meta added

File: pkgs/tools/X11/xdg-open-with-portal/default.nix [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    fi
+
+    ${glib}/bin/gdbus call --session \
+      --dest org.freedesktop.portal.Desktop \
+      --object-path /org/freedesktop/portal/desktop \
+      --method org.freedesktop.portal.OpenURI.OpenURI \
+      --timeout 5 \
+      "" "$targetFile" {}
+  fi
+''
```
Comment (LunNova) [inline pkgs/tools/X11/xdg-open-with-portal/default.nix:37]: - [ ] FIXME: needs meta added

File: nixos/modules/config/xdg/xdg-open.nix [hunk 3]
```diff
@@ -0,0 +1,44 @@
+{ config, pkgs, lib, ... }:
+let
+  cfg = config.xdg.xdg-open;
+  implementations = {
```
Comment (LunNova) [inline nixos/modules/config/xdg/xdg-open.nix:4]: TODO: make this support custom impl packages, this approach isn't great

**Comment by FliegendeWurst** [comment]
This solves #160923 for Trilium Notes (wrapped electron application) :+1:

**Review by Atemu — COMMENTED** [review]
I don't think duplicating upstream code is a good idea.

We should rather patch the regular xdg-open to handle our FHSEnvs the same as other container runtimes (i.e. flatpak).  
We should probably even propose that upstream. Perhaps not in specific to NixOS but for all containerised apps.

Alternatively, we could also just pretend our FHSEnv'd apps are flatpaks.

**Comment by LunNova** [comment]
> I don't think duplicating upstream code is a good idea.

It isn't really duplicating code? It's a small shell script that always uses the portal, and it's not just for FHS envs, it's also for running under simple wrapper scripts or any other context.

It doesn't use any xdg-open code and makes the portal responsible for things 100% of the time.

I could make a small repo for it under github.com/LunNova and package it from github instead of including it inline.

**Comment by Atemu** [comment]
I don't see how we're not duplicating upstream code here? We've now got a shell script that is supposed to do the same as upstream xdg-open and even uses the same mechanism internally but we need to maintain it separately.

The location of the code is not the problem, the fact that we've got a separate implementation of xdg-open at all is.

I think we should instead have a minimal patch that makes xdg-open detect our fhsenvs and treat them the same as flatpak. That's very simply to do. Perhaps even propose that patch upstream so we don't have to maintain it downstream.

Building a new implementation of xdg-open and then globally(!) switching it out to fix a bug in fhsenvs is not sustainable in my eyes.

**Comment by LunNova** [comment]
In general being able to replace the version of xdg-open on your system with something not from xdg-utils is useful. There's also [handlr](https://github.com/chmln/handlr) and [xdg_open_ng](https://git.syndicate-lang.org/ehmry/xdg_open_ng), it's not unusual to want to swap your xdg-open implementation. Having a shell script called 'xdg-open' on path is the API that programs expect, and xdg-utils is not the only source of an implementation of it in the wild.

In my mind "xdg-open-with-portal" is a package that provides the `xdg-open` API and its implementation always uses dbus to ask the xdg portal to open an app. It's very simple, and doesn't contain most of the functionality of xdg-utils xdg-open, because that's now the portal's job. Since it assumes a portal exists it can be very minimal. It isn't derived from xdg-utils or a copy/paste job from upstream, it's a simple shell script I wrote for this use case. I think it fits into the same category as `handlr` or `xdg_open_ng`.

Separately, it'd be good to get xdg-utils to support some env var to tell it to always use the portal to open things, and to add support for opening files through the portal, but I don't think that should prevent adding a nixos module which controls which implementation of xdg-open you, or that it should prevent having a simple shell script impl of xdg-open that uses the portal.

**Comment by LunNova** [comment]
@Atemu how do you feel about including https://github.com/freedesktop/xdg-utils/pull/12 in nixpkgs in a separate PR before it gets merged?

xdg-utils seem very slow to respond to PRs.

**Comment by Atemu** [comment]
I'd make it a patch behind a flag.

**Comment by bb010g** [comment]
> Raised [freedesktop/xdg-utils/pull/12](https://github.com/freedesktop/xdg-utils/pull/12) upstream to add an env var to tell `xdg-open` to always use the portal.

This repository no longer exists, and I don't see a matching MR on <https://gitlab.freedesktop.org/xdg/xdg-utils>.

**Comment by LunNova** [comment]
@bb010g it was this commit: https://github.com/LunNova/xdg-utils/commit/123e8cedaff6c4708c82ed8b263d497fd00c8d5c

**Comment by LunNova** [comment]
Closed in favor of #197118

**Pull request closed without merging**